x11-utils.c

```diff
@@ -0,0 +1,227 @@
+#include <X11/extensions/XInput.h>
+#include <X11/extensions/XInput2.h>
+#include <X11/extensions/dpms.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "x11-utils.h"
+#include "mce-log.h"
+
+static Atom x11_atom_touchscreen = None;
+static Atom x11_atom_device_enabled = None;
+static Atom x11_atom_device_enabled_type = None;
+static int x11_atom_device_enabled_format = 0;
+
+static Display* x11_get_display(void)
+{
+		Display* dpy = XOpenDisplay(NULL);
+		if (dpy == NULL) {
+			dpy = XOpenDisplay(":0.0");
+		}
+
+		if (dpy == NULL) {
+			mce_log(LL_INFO, "%s: unable to open display", __func__);
+		}
+		return dpy;
+}
+
+bool x11_set_input_device_enabled(Display* dpy, const XIDeviceInfo devinfo, const bool enable)
+{
+	if (dpy == NULL) {
+		mce_log(LL_WARN, "%s: XIDeviceInfo needs to be from same Display", __func__);
+		return false;
+	}
+	if (x11_atom_device_enabled == None)
+		x11_atom_device_enabled = XInternAtom(dpy, "Device Enabled", False);
+	if (x11_atom_device_enabled == None) {
+		mce_log(LL_WARN, "%s: unable to obtain X11 Atoms", __func__);
+		return false;
+	}
+	
+	if ((x11_atom_device_enabled_type == None) || (x11_atom_device_enabled_format == 0)) {
+		unsigned long ignore_bytes_after, ignore_nitems;
+		unsigned char *ignore_data = NULL;
+
+		if (XIGetProperty(dpy, devinfo.deviceid, x11_atom_device_enabled, 0, 0, False, AnyPropertyType, &x11_atom_device_enabled_type, &x11_atom_device_enabled_format, &ignore_nitems, &ignore_bytes_after, &ignore_data)) {
+			mce_log(LL_WARN, "%s: unable to obtain X11 Device Enabled property atom type", __func__);
+			return false;
+		} else {
+			XFree(ignore_data);
+		}
+	}
+	
+	XIChangeProperty(dpy, devinfo.deviceid, x11_atom_device_enabled,
+			x11_atom_device_enabled_type,
+			x11_atom_device_enabled_format, PropModeReplace, (unsigned char*)&enable, 1);
+	return true;
+}
+
+bool x11_set_all_input_devices_enabled(Display* dpy, const bool enable)
+{
```

Comment:
Display *dpy. Also - why the 'const' on that bool?

Reply:
Why not const on that bool? we dont intend modify it, this is best practice.

Reply:
I was just wondering how it is better, given that the variable is passed by value anyway.

Reply:
Well it conveys intention: this variable is not supposed to be modified that allows the compiler to provide better warnings when a future programmer mistypes (classic = vs == typo). It also helps the compiler to optimize away the copy in some instances, but these instances are not relevant here.